jax/custom_derivatives.py

```diff
@@ -837,3 +838,107 @@ def tree_flatten(self):
   def tree_unflatten(cls, aux, consts):
     jaxpr, in_tree, out_tree = aux
     return cls(jaxpr, in_tree, out_tree, consts)
+
+
+def closure_convert(fun, example_args):
+  """Closure conversion utility, for use with higher-order custom derivatives.
+
+  To define custom derivatives such as with ``jax.custom_vjp(f)``, the target
+  function ``f`` must take, as formal arguments, all values involved in
+  differentiation. If ``f`` is a higher-order function, in that it accepts as an
+  argument a Python function ``g``, then values stored away in ``g``'s closure
+  will not be visible to the custom derivative rules, and attempts at AD
+  involving these values will fail. One way around this is to convert the
+  closure by extracting these values, and to pass them as explicit formal
+  arguments across the custom derivative boundary. This utility carries out that
+  conversion. More precisely, it closure-converts the function ``fun``
+  specialized to the types of the arguments given in ``example_args``.
+
+  When we refer here to "values in the closure" of ``fun``, we do not mean the
+  values that are captured by Python directly when ``fun`` is defined (e.g. the
+  Python objects in ``fun.__closure__``, if the attribute exists). Rather, we
+  mean values encountered during the execution of ``fun`` on ``example_args``
+  that determine its output. This may include, for instance, arrays captured
+  transitively in Python closures, i.e. in the Python closure of functions
+  called by ``fun``, the closures of the functions that they call, and so forth.
+
+  The function ``fun`` must be a pure function.
+
+  Example usage::
```

Comment:
This is a great example. Could you add a test that basically copy-pastes it? That way we get a test almost for free, and we're confident the docstring is accurate! (The test need not do any actual optimization; its implementation could be trivial perhaps.)

Reply:
Good call. Done.